lib/improver/convection.py

```diff
@@ -0,0 +1,207 @@
+        2. Threshold the input cube using the higher_threshold and
+           apply neighbourhood processing.
+        3. Calculate the convective ratio by:
+           higher_threshold_cube / lower_threshold_cube.
+           For example, the higher_threshold might be 5 mm/hr, whilst the
+           lower_threshold might be 0 mm/hr.
+
+        The Convective Ratio can have the following values:
+            * A non-zero fractional value, indicating that both the higher
+              and lower thresholds were exceeded.
+            * A zero value, if the lower threshold was exceeded, whilst the
+              higher threshold was not exceeded.
+            * A NaN value (np.nan), if neither the higher or lower thresholds
+              were exceeded, and therefore a NaN value results from the
+              0/0 division.
+
+        Args:
+            cube : Iris.cube.Cube
+                The cube from which the convective ratio will be calculated.
+
+        Returns:
+            convective_ratio : Iris.cube.Cube
+                Cube containing the convective ratio
+        """
+        threshold_dict = {}
+
+        for threshold in [self.lower_threshold, self.higher_threshold]:
+            threshold_cube = (
+                BasicThreshold(
+                    threshold, fuzzy_factor=self.fuzzy_factor,
+                    below_thresh_ok=self.below_thresh_ok).process(cube.copy()))
+            neighbourhooded_cube = NeighbourhoodProcessing(
+                self.neighbourhood_method, self.radii_in_km,
+                lead_times=self.lead_times,
+                unweighted_mode=self.unweighted_mode,
+                ens_factor=self.ens_factor).process(threshold_cube)
+            threshold_dict[threshold] = neighbourhooded_cube
+
+        # Ignore runtime warnings from divide by 0 errors.
+        with np.errstate(invalid='ignore'):
+            convective_ratio = (
+                threshold_dict[self.higher_threshold] /
+                threshold_dict[self.lower_threshold])
+        if np.sum(np.isinf(convective_ratio.data)) > 0.0:
+            msg = ("A value of infinity was found for the "
+                   "convective ratio: {}.\n"
+                   "This value is not plausible as the fraction above the "
+                   "higher threshold must be lower than the fraction "
+                   "above the lower threshold.").format(
+                       convective_ratio.data)
+            raise ValueError(msg)
+        convective_ratio.long_name = "convective_ratio"
+        return convective_ratio
+
+    def process(self, cube):
+        """
+        Calculate the convective ratio either for the underlying field e.g.
+        precipitation rate, or using the differences between adjacent grid
+        squares.
+
+        Args:
+            cube : Iris.cube.Cube
+                The cube from which the convective ratio will be calculated.
+
+        Returns:
+            convective_ratios : Iris.cube.CubeList
```

Comment:
If using the differences between adjacent squares, how is the output returned different compared to when these differences are not used? For instance will there be two cubes (x, y) rather than one? Should state any differences here.